scripts/vehicle_node.py

```diff
@@ -0,0 +1,304 @@
+        w = self.resolution
+        grid_map = (h,w)
+        gaussian_array = np.random.normal(self.mu, self.sigma, grid_map)
+
+        #Filter the array to smoothen the variation of the noise
+        gaussian_array = gaussian_filter(gaussian_array, self.sigma_filt)
+
+        #Create an list and fill the list with x, y values and z the values will be the filtered gaussian noise
+        # TODO perform these calculations without for loops
+        incr_x = 0
+        incr_y = 0
+        cloud = []
+        for x in np.linspace((-self.height), (self.height), self.resolution):
+            for y in np.linspace((-self.width), (self.width), self.resolution):
+                innerlist = []
+                innerlist.append(x)
+                innerlist.append(y)
+                innerlist.append(gaussian_array[incr_x][incr_y])
+                cloud.append(innerlist)
+                incr_y+=1
+            incr_x+= 1
+            incr_y = 0
+
+        return cloud
+
+
+    def velocity_cmd_callback(self, data):
+        
+        """
+        Updates the most recent command velocity to a twist variable
+        """
+        self.lock.acquire()
+        try:
+            self.twist = data
+        finally:
+            self.lock.release()
+
+
+    def update_position(self, event):
+
+        """
+        Computes the pose of the vehicle within a given footprint height and width 
+        """
+
+        #Create a copy of the most recent stored twist data to perform calculations with
+        self.lock.acquire()
+        try:
+            velocity_data = copy.deepcopy(self.twist)
+        finally:
+            self.lock.release()
+
+        #time elapsed since last update position call
+        if hasattr(event, 'last_real'):
+            if event.last_real is None:
+                time = rospy.Duration(0.05)
+            else:
+                time = event.current_real - event.last_real
+        
+        time = time.to_sec()
+
+        #Calculate angle turned in the given time using omega = theta/time
+        #TODO Currently it yaws, then moves forward. Revist calculation to do both simultaneously
+        angle = velocity_data.angular.z*time
+
+        #Calculate distance travelled in the given time using linear velocity = arc distance/time
+        distance = velocity_data.linear.x*time
+
+        #Calculate yaw, pitch and roll of the robot (pitch and roll currently not calculated)
+        vehicle_roll = 0.0
+        vehicle_pitch = 0.0
+        self.vehicle_yaw += angle
+
+        #Calculate vehicle x, y, z position coordinates 
+        self.pose.position.x += (distance)*cos(self.vehicle_yaw)
+        self.pose.position.y += (distance)*sin(self.vehicle_yaw)
+    
+        #Convert the point cloud to an array for faster access
+        myarray = np.asarray(self.cloud)
+
+        points = []
+        values = []
+
+        #Find all the points within the point cloud array that lie within the footprint of the vehicle.
```

Comment:
The point cloud should be used only for visualization.  For all calculations below, you should directly use a 2 dimensions numpy array, since that will make the calculations much more efficient.